### ISSUE TITLE
A bug in the fuzzy hashing caused by bit width

### ISSUE THREAD

**Issue by fridex** [thoth-station/storages#2098]
**Describe the bug**

The input to the fuzzy hashing algorithm does not behave on the binary level as expected, see:

https://github.com/thoth-station/storages/blob/989b625cfd788da1b950cd458763598655234b2d/thoth/storages/graph/postgres.py#L3729-L3733

This causes different size allocation for different integer sizes. This means that different integers will have allocated different size causing the fuzzy hashing not working properly. Each integer should be assigned a constant width (with zero padding) so that fuzzy hashing can recognize differences across hashing for inputs with different sizes.

**To Reproduce**

```python
>>> a = 1
>>> a.bit_length()
1
>>> a = 100
>>> a.bit_length()
7
```

**Expected behavior**

The hashing should be done on a fixed size integers.

**Comment by goern** [comment]
@fridex what do we need to do about this?

**Comment by fridex** [comment]
> @fridex what do we need to do about this?

Use fixed with for integers when hashing a list of hashes. Otherwise fuzzy hashing will not produce reasonable results.

**Comment by goern** [comment]
ja, and do we know all the places where we need to fix the width? is that something that a unit test should cover?

**Comment by fridex** [comment]
> ja, and do we know all the places where we need to fix the width? is that something that a unit test should cover?

Yes, I think so.

**Comment by KPostOffice** [comment]
To solve this we need to extend _fuzzy_hash method to accept width and come up with good numbers based on a reasonable `MAX` for each id field and we will also need to update the hashes in the database right?

**Comment by fridex** [comment]
> To solve this we need to extend _fuzzy_hash method to accept width and come up with good numbers based on a reasonable `MAX` for each id field and we will also need to update the hashes in the database right?

Yes, that's correct. 👍🏻 The old entries can stay as they are. As we have datetime/version info and old stacks are not that relevant to us, we can filter them (not to spend too much time on this work).

**Comment by goern** [comment]
@fridex @KPostOffice can one of you take this and work fix it?

**Comment by fridex** [comment]
Let's have this low priority. In the worst case, we will have bad hashes for an initial set of stacks kept. These hashes can be recomputed later on once a fix is done. Also, we do not have any use of these hashes planned as of now.

/priority important-longterm
